### PR TITLE
[55] Implement report UI

### PR DESCRIPTION
**Pull request by pisuttun**
## What did you do

- Implement Report page

- Add NotiContext (will write more on notion after this PR got merged)

image

## Demo

https://dev.modela.miello.dev/report/1 //login as actor first

### CODE REVIEW

File: apps/web/src/modules/report/hooks/useReport/index.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import { zodResolver } from '@hookform/resolvers/zod'
+import { GetJobDto } from '@modela/dtos'
+import { useNoti } from 'common/context/NotiContext'
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/index.ts:3]: I'm quite not recommend to abbreviate from `notification`

File: apps/web/src/modules/report/hooks/useReport/index.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    },
+  })
+
+  const handleSuccess: SubmitHandler<IReportSchemaType> = useCallback(
+    async ({ ...data }) => {
+      try {
+        const postBody = {
+          reason: data.description,
+        }
+        setLoading(true)
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/index.ts:31]:
```suggestion
      setLoading(true)
      try {
        const postBody = {
          reason: data.description,
        }
```

File: apps/web/src/modules/report/hooks/useReport/index.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  const { control, handleSubmit } = useForm<IReportSchemaType>({
+    criteriaMode: 'all',
+    resolver: zodResolver(ReportSchema),
+    defaultValues: {
+      description: '',
+    },
+  })
+
+  const handleSuccess: SubmitHandler<IReportSchemaType> = useCallback(
+    async ({ ...data }) => {
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/index.ts:26]: Do we have any concern that making us to use spread operator ?

File: apps/web/src/modules/report/hooks/useReport/index.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        console.log(err)
+      } finally {
+        setLoading(false)
+      }
+    },
+    [router],
+  )
+
+  const handleClickSubmit: FormEventHandler<HTMLFormElement> =
+    handleSubmit(handleSuccess)
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/index.ts:48]: useMemo might help performance

File: apps/web/src/modules/report/hooks/useReport/schema.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { z } from 'zod'
+
+const ReportSchema = z.object({
+  description: z
+    .string({ required_error: 'กรุณากรอกรายละเอียดปัญหา' })
+    .min(1, 'กรุณากรอกรายละเอียดปัญหา'),
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/schema.ts:6]: Not neccessary ?

File: apps/web/src/modules/report/pages/index.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+      <Button
+        sx={{ borderRadius: '12px', alignSelf: 'center' }}
+        size="large"
+        variant="contained"
+        type="submit"
+        startIcon={loading && <CircularProgress size={20} />}
+      >
+        แจ้งปัญหา
+      </Button>
```
Comment (miello) [inline apps/web/src/modules/report/pages/index.tsx:62]: Please disable submit button when we click submit for preventing click twice

File: apps/web/src/common/context/NotiContext/index.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  const [type, setType] = React.useState<NotiType>('success')
+
+  const handleClose = () => setShowNoti(false)
+
+  const displayNoti = (message: string, type: NotiType, duration?: number) => {
+    setNotiMessage(message)
+    setType(type)
+    duration && setDuration(duration)
+    setShowNoti(true)
+  }
```
Comment (miello) [inline apps/web/src/common/context/NotiContext/index.tsx:24]: Consider using `useCallback` here

File: apps/web/src/common/context/NotiContext/index.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  const [showNoti, setShowNoti] = React.useState(false)
+  const [duration, setDuration] = React.useState(3000)
+  const [type, setType] = React.useState<NotiType>('success')
+
+  const handleClose = () => setShowNoti(false)
+
+  const displayNoti = (message: string, type: NotiType, duration?: number) => {
+    setNotiMessage(message)
+    setType(type)
+    duration && setDuration(duration)
```
Comment (miello) [inline apps/web/src/common/context/NotiContext/index.tsx:22]:
```suggestion
   if (duration) setDuration(duration)
```

File: apps/web/src/common/context/NotiContext/index.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+export const useNoti = () => React.useContext(NotiContext)
+
+export const NotiProvider = ({ children }: React.PropsWithChildren<{}>) => {
+  const [notiMessage, setNotiMessage] = React.useState('')
+  const [showNoti, setShowNoti] = React.useState(false)
+  const [duration, setDuration] = React.useState(3000)
+  const [type, setType] = React.useState<NotiType>('success')
+
+  const handleClose = () => setShowNoti(false)
```
Comment (miello) [inline apps/web/src/common/context/NotiContext/index.tsx:17]: Consider using `useCallback` here

File: apps/web/src/modules/report/pages/index.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        render={(prop) => (
+          <TextField
+            fullWidth
+            multiline
+            rows={3}
+            {...prop.field}
+            error={prop.fieldState.invalid}
+            helperText={prop.fieldState.error?.message}
+            placeholder="เหตุผล"
+          />
```
Comment (miello) [inline apps/web/src/modules/report/pages/index.tsx:50]:
```suggestion
          <TextField
            fullWidth
            multiline
            rows={3}
            {...prop.field}
            inputRef={prop.field.ref}
            error={prop.fieldState.invalid}
            helperText={prop.fieldState.error?.message}
            placeholder="เหตุผล"
          />
```

File: apps/web/src/modules/report/hooks/useReport/index.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      description: '',
+    },
+  })
+
+  const handleSuccess: SubmitHandler<IReportSchemaType> = useCallback(
+    async ({ ...data }) => {
+      try {
+        const postBody = {
+          reason: data.description,
+        }
```
Comment (miello) [inline apps/web/src/modules/report/hooks/useReport/index.ts:30]: Please specify dtos type from `@model/dtos`

**Review by miello — CHANGES_REQUESTED** [review]

File: apps/web/src/modules/report/hooks/useReport/index.ts (hunk 3 above)
Comment (pisuttun) [reply apps/web/src/modules/report/hooks/useReport/index.ts:26]: idk, I just copy this from your signup form 😂. should I change this?

Comment (miello) [reply apps/web/src/modules/report/hooks/useReport/index.ts:26]:
```suggestion
    async (data) => {
```

Comment (miello) [reply apps/web/src/modules/report/hooks/useReport/index.ts:26]: Or maybe 
```suggestion
    async ({ description }) => {
```
then change from `data.description` to `description`

**Review by miello — APPROVED** [review]

**Pull request merged**